superduperdb/ext/llm/vllm.py

```diff
@@ -0,0 +1,169 @@
+import asyncio
+import dataclasses as dc
+from typing import Any, List, Optional
+
+import aiohttp
+import requests
+
+from superduperdb import logging
+from superduperdb.ext.llm.base import BaseLLMAPI, BaseLLMModel, BaseOpenAI
+
+VLLM_INFERENCE_PARAMETERS_LIST = [
+    "n",
+    "best_of",
+    "presence_penalty",
+    "frequency_penalty",
+    "repetition_penalty",
+    "temperature",
+    "top_p",
+    "top_k",
+    "min_p",
+    "use_beam_search",
+    "length_penalty",
+    "early_stopping",
+    "stop",
+    "stop_token_ids",
+    "include_stop_str_in_output",
+    "ignore_eos",
+    "max_tokens",
+    "logprobs",
+    "prompt_logprobs",
+    "skip_special_tokens",
+    "spaces_between_special_tokens",
+    "logits_processors",
+]
+
+
+@dc.dataclass
+class VllmOpenAI(BaseOpenAI):
+    """
+    Wrapper for requesting the vLLM API service
+    (OpenAI format, started by vllm.entrypoints.openai.api_server)
+
+    {parent_doc}
+    """
+
+    openai_api_key: Optional[str] = "EMPTY"
+
+    __doc__ = __doc__.format(parent_doc=BaseOpenAI.__doc__)
+
+
+@dc.dataclass
+class VllmAPI(BaseLLMAPI):
+    """
+    Wrapper for requesting the vLLM API service
+    (API Server format, started by vllm.entrypoints.api_server)
+    {parent_doc}
+    """
+
+    __doc__ = __doc__.format(parent_doc=BaseLLMAPI.__doc__)
+
+    def _generate(self, prompt: str, **kwargs) -> str:
+        """
+        Batch generate text from a prompt.
+        """
+        post_data = self.build_post_data(prompt, **kwargs)
+        response = requests.post(self.api_url, json=post_data)
+        return response.json()["text"][0]
+
+    def _batch_generate(self, prompts: List[str], **kwargs: Any) -> List[str]:
+        """
+        Use asyncio to batch generate text from a list of prompts.
+        """
+        return asyncio.run(self._async_batch_generate(prompts, **kwargs))
+
+    async def _async_generate(self, session, semaphore, prompt: str, **kwargs) -> str:
+        post_data = self.build_post_data(prompt, **kwargs)
+        async with semaphore:
+            try:
+                async with session.post(self.api_url, json=post_data) as response:
+                    response_json = await response.json()
+                    return response_json["text"][0]
+            except aiohttp.ClientError as e:
+                logging.error(f"HTTP request failed: {e}. Prompt: {prompt}")
+                return ""
+
+    async def _async_batch_generate(self, prompts: List[str], **kwargs) -> List[str]:
+        async with aiohttp.ClientSession() as session:
+            semaphore = asyncio.Semaphore(self.max_batch_size or len(prompts))
+            tasks = [
+                self._async_generate(session, semaphore, prompt, **kwargs)
+                for prompt in prompts
+            ]
+            return await asyncio.gather(*tasks)
+
+    def build_post_data(self, prompt: str, **kwargs: dict[str, Any]) -> dict[str, Any]:
```

Comment:
Please explain this.

Reply:
Combined with the prompt to wrap the request data of the vLLM API, it supports model predefined parameters and parameters passed when calling, and finally performs parameter filtering (if unnecessary parameters are passed in, an error will occur, such as the context introduced by SuperDuperDB, or others).